### PR TITLE
Allow setting of a maximum page size to be used with DynamoDB reads to attempt to allow better control of throughputs

### PR DESCRIPTION
**Pull request by clicktravel-craig**
This commit adds the ability to set a maximum page size that DynamoDB will use when reading data, this will allow me to run tests on UAT to confirm whether changing the max page size will have the desired impact on reducing read throughputs.

Signed-off-by: Craig Handley <craig.handley@clicktravel.com>

the-cheddar-chapter@clicktravel.com - please review

### CODE REVIEW

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDbTemplate.java [hunk 1]
```diff
@@ -412,21 +415,38 @@ public void delete(final Item item, final PersistenceExceptionHandler<?>... pers
 
     @Override
     public <T extends Item> Collection<T> fetch(final Query query, final Class<T> itemClass) {
+        return fetch(query, itemClass, null);
+    }
+
+    @Override
+    public <T extends Item> Collection<T> fetch(final Query query, final Class<T> itemClass,
+            final Integer maxPageSize) {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDbTemplate.java]: Always prefer `int` over `Integer` unless `null` has a meaning in the given context.

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDbTemplate.java [hunk 2]
```diff
@@ -412,21 +415,38 @@ public void delete(final Item item, final PersistenceExceptionHandler<?>... pers
...
         final long elapsedTimeMillis = System.currentTimeMillis() - startTimeMillis;
+
         logger.debug("Database fetch executed in " + elapsedTimeMillis + "ms. Query:[" + query + "]");
+
         return result;
     }
 
-    private <T extends Item> Collection<T> executeQuery(final AttributeQuery query, final Class<T> itemClass) {
+    private <T extends Item> Collection<T> executeQuery(final AttributeQuery query, final Class<T> itemClass,
+            final Integer maxPageSize) {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDbTemplate.java]: See comment above about `Integer`

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDbTemplate.java [hunk 3]
```diff
@@ -463,39 +483,56 @@ public void delete(final Item item, final PersistenceExceptionHandler<?>... pers
...
             do {
                 final String queryAttributeName = query.getAttributeName();
                 final PrimaryKeyDefinition primaryKeyDefinition = itemConfiguration.primaryKeyDefinition();
                 final String primaryKeyPropertyName = primaryKeyDefinition.propertyName();
                 final boolean isPrimaryKeyQuery = queryAttributeName.equals(primaryKeyPropertyName);
+
                 final QueryRequest queryRequest = new QueryRequest().withTableName(tableName)
                         .withKeyConditions(conditions).withExclusiveStartKey(lastEvaluatedKey);
+
+                if (maxPageSize != null) {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDbTemplate.java]: `maxPageSize > 0` is a more fluent API

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDbTemplate.java [hunk 4]
```diff
@@ -463,39 +483,56 @@ public void delete(final Item item, final PersistenceExceptionHandler<?>... pers
...
             } while (lastEvaluatedKey != null);
 
         } else {
             logger.debug("Performing table scan with query: " + query);
+
             do {
                 final ScanRequest scanRequest = new ScanRequest().withTableName(tableName).withScanFilter(conditions)
                         .withExclusiveStartKey(lastEvaluatedKey);
+
+                if (maxPageSize != null) {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDbTemplate.java]: See comment about fluent API

File: cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/integration/DynamoDbTemplateIntegrationTest.java [hunk 5]
```diff
@@ -426,6 +426,26 @@ public void shouldFetch_withAttributeQuery() throws Exception {
...
+        final StubItem createdItem = dataGenerator.createStubItem();
+        final String stringProperty = createdItem.getStringProperty();
+        final StubItem stubItem = new StubItem();
+        stubItem.setStringProperty(stringProperty);
+        final Query query = new AttributeQuery(STRING_PROPERTY, new Condition(Operators.EQUALS, stringProperty));
+        final DynamoDbTemplate dynamoDbTemplate = new DynamoDbTemplate(databaseSchemaHolder);
+        dynamoDbTemplate.initialize(amazonDynamoDbClient);
+
+        // When
+        final Collection<StubItem> itemResults = dynamoDbTemplate.fetch(query, StubItem.class, 10);
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/integration/DynamoDbTemplateIntegrationTest.java]: Prefer `randomInt(11)` (returned values: 0 - 10)

File: cheddar/cheddar-integration-mocks/src/main/java/com/clicktravel/infrastructure/persistence/inmemory/database/InMemoryDatabaseTemplate.java [hunk 6]
```diff
@@ -209,6 +211,12 @@ boolean hasMatchingUniqueConstraint(final Item item, final String propertyName,
     }
 
     @Override
+    public <T extends Item> Collection<T> fetch(final Query query, final Class<T> itemClass,
+            final Integer maxPageSize) {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-mocks/src/main/java/com/clicktravel/infrastructure/persistence/inmemory/database/InMemoryDatabaseTemplate.java]: `int`

**Comment by clicktravel-basharat** [comment]
Round one done.

**Comment by clicktravel-craig** [comment]
@clicktravel-basharat comments applied

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDbTemplate.java [hunk 7]
```diff
@@ -463,39 +474,56 @@ public void delete(final Item item, final PersistenceExceptionHandler<?>... pers
...
             do {
                 final String queryAttributeName = query.getAttributeName();
                 final PrimaryKeyDefinition primaryKeyDefinition = itemConfiguration.primaryKeyDefinition();
                 final String primaryKeyPropertyName = primaryKeyDefinition.propertyName();
                 final boolean isPrimaryKeyQuery = queryAttributeName.equals(primaryKeyPropertyName);
+
                 final QueryRequest queryRequest = new QueryRequest().withTableName(tableName)
                         .withKeyConditions(conditions).withExclusiveStartKey(lastEvaluatedKey);
+
+                if (maxPageSizeOverridden(maxPageSize)) {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDbTemplate.java]: `maxPageSize > 0` would be clearer. Perhaps `limit` is better than `maxPageSize`?

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java [hunk 8]
```diff
@@ -348,9 +367,22 @@ private QuerySpec generateQuerySpec(final AttributeQuery query) {
...
         }
+
+        if (maxPageSizeOverridden(maxPageSize)) {
+            scanSpec = scanSpec.withMaxPageSize(maxPageSize);
+        }
+
         return scanSpec;
     }
 
+    private boolean maxPageSizeOverridden(final int maxPageSize) {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java]: See above comment related to similar code.

**Comment by clicktravel-basharat** [comment]
Round #2 done.

**Comment by clicktravel-craig** [comment]
@clicktravel-basharat limit did cross my mind but I thought limit seemed to relate more to the overall number of records that would be returned rather than the maximum number of records in the current fetch. Looking through AWS docs they refer to it as page size also http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/QueryAndScanGuidelines.html
Let me know one way or the other which you think would be better

**Comment by clicktravel-basharat** [comment]
After more thought, the max page size should not feature on the method as the maximum page size is an internal setting which does not affect the overall result of the `query` method. It seems a template-wide config rather than per-request param.

I think what would be most appropriate so to have the `AbstractDynamoDbTemplate` accept some config which can set the max page size.

Or do we indeed have a concrete need to set it on a per-request basis, that cannot be achieved by wiring in multiple DB templates?

**Comment by clicktravel-craig** [comment]
@clicktravel-basharat Can I suggest that we should just get it in as is. The main reason I need this in is so that I can actually test it in the UAT environment, as to be honest this may not even work at all, so probably best not to spend too much time on it now. If we get it in and it has the desired effect, then I think we should refactor to apply in the best way.

**Pull request closed without merging**